sdk/kiosk/src/client/tp-manager.ts

```diff
@@ -0,0 +1,371 @@
+	kioskLockRuleAddress,
+	personalKioskAddress,
+	royaltyRuleAddress,
+} from '../constants';
+
+export class TransferPolicyManager {
+	client: SuiClient;
+	network: Network;
+	policyId?: ObjectArgument;
+	policyCap?: ObjectArgument;
+	itemType?: string;
+
+	constructor(options: KioskClientOptions) {
+		this.client = options.client;
+		this.network = options.network;
+	}
+
+	/**
+	 * A function to create a new transfer policy.
+	 * Checks if there's already an existing transfer policy to prevent
+	 * double transfer polciy mistakes.
+	 * There's an optional `skipCheck` flag that will just create the policy
+	 * without checking
+	 *
+	 * @param tx The Transaction Block
+	 * @param itemType The Type (`T`) for which we're creating the transfer policy.
+	 * @param publisher The Publisher Object Id.
+	 * @param address Address to save the `TransferPolicyCap` object to.
+	 */
+	async createAndShare(
+		tx: TransactionBlock,
+		itemType: string,
+		publisher: ObjectArgument,
+		address: string,
+		options?: {
+			skipCheck?: boolean;
+		},
+	) {
+		if (!options || !options.skipCheck) {
+			let policies = await queryTransferPolicy(this.client, itemType);
+			if (policies.length > 0) throw new Error("There's already transfer policy for this Type.");
+		}
+		const cap = createTransferPolicy(tx, itemType, publisher);
+		tx.transferObjects([cap], tx.pure(address, 'address'));
+	}
+
+	/**
+	 * A convenient function to create a Transfer Policy and attach some rules
+	 * before sharing it (so you can prepare it in a single PTB)
+	 * @param tx The Transaction Block
+	 * @param itemType The Type (`T`) for which we're creating the transfer policy.
+	 * @param publisher The Publisher Object Id.
+	 * @param address Address to save the `TransferPolicyCap` object to.
+	 */
+	async create(
+		tx: TransactionBlock,
+		itemType: string,
+		publisher: ObjectArgument,
+		options?: {
+			skipCheck?: boolean;
+		},
+	): Promise<TransferPolicyManager> {
+		if (!options || !options.skipCheck) {
+			let policies = await queryTransferPolicy(this.client, itemType);
+			if (policies.length > 0) throw new Error("There's already transfer policy for this Type.");
+		}
+		const [transferPolicy, cap] = createTransferPolicyWithoutSharing(tx, itemType, publisher);
+
+		this.setPolicy(transferPolicy, cap, itemType); // sets the client's TP to the newly created one.
+		return this;
+	}
+
+	/**
+	 * This can be called after calling the `create` function to share the `TransferPolicy`,
+	 * and transfer the `TransferPolicyCap` to the specified address
+	 */
+	shareAndTransferCap(tx: TransactionBlock, address: string) {
+		if (!this.itemType || !this.policyCap || !this.policyId)
+			throw new Error('This function can only be called after `transferPolicyManager.create`');
+
+		shareTransferPolicy(tx, this.itemType, this.policyId as TransactionArgument);
+		tx.transferObjects([this.policyCap as TransactionArgument], tx.pure(address, 'address'));
+
+		this.resetPolicy();
+	}
+
+	resetPolicy() {
```

Comment:
This API feels wrong to me, we shouldn't be building up state and resetting it.

Reply:
@Jordan-Mysten I was thinking of making`resetPolicy` private. The reason it needs reset is that you can build this in two ways, either by passing an existing object, or [by creating it and building upon it on the same PTB](https://sui-typescript-docs-git-ml-kiosk-sdk-v2-docs-mysten-labs.vercel.app/kiosk/transfer-policy-manager/using-the-manager#creation--attachment-of-rules-in-one-ptb). When you do it by the latter, you can't keep using it. You need to sign, execute, and then pass it as an existing object.